<!doctype html>
<html>
	<head>
	<meta charset="utf-8">
	<meta http-equiv="X-UA-Compatible" content="chrome=1">
	<title>RNFtools</title>

	<link rel="stylesheet" href="stylesheets/styles.css">
	<link rel="stylesheet" href="stylesheets/pygment_trac.css">
	<script src="javascripts/scale.fix.js"></script>
	<meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
	<!--[if lt IE 9]>
	<script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
	<![endif]-->

	<style type="text/css">
		/* section.main_section a{color:#0000dd;} */
		header{}
		footer {margin:0px;padding:0px;width:250px;text-align:center;}
		footer img {border-width: 0px;height:90px;}
	</style>
	</head>
	<body>
	<div class="wrapper">
		<header>
			<h1 class="header">RNFtools</h1>

			<ul>
				<li class="download">
					<a href="http://rnftools.rtfd.org">Documentation</a>
				</li>
				<li class="download">
					<a href="http://karel-brinda.github.io/rnf-spec/RNF.pdf">RNF spec.</a>
				</li>
				<li class="download">
					<a href="http://dx.doi.org/10.1093/bioinformatics/btv524">Publication</a>
				</li>
				<li class="download">
					<a href="http://groups.google.com/group/rnftools/">Mailing list</a>
				</li>
				<br />
				<li class="download">
					<a href="https://github.com/karel-brinda/rnftools">View On GitHub</a>
				</li>
				<li class="download">
					<a href="https://github.com/karel-brinda/rnftools/tarball/master">Download TAR</a>
				</li>
			</ul>

		<!--
		<p class="header">This project is maintained by <a class="header name" href="https://github.com/karel-brinda">karel-brinda</a></p>
		-->


		</header>
		<section class="main_section">

			<h1>Read Naming Format and RNFtools</h1>

<!--
	========================================================
-->
<h2>RNF</h2>

<p>
	<strong>RNF</strong> (Read Naming Format, <a href="http://karel-brinda.github.io/rnf-spec/RNF.pdf">specification</a>)
	is a generic format for assigning names to simulated Next-Generation Sequencing (NGS) reads. It provides a unified
	way of encoding genomic coordinates of reads,
	no matter which specific sequencing technology is simulated or which read simulator employed.
	Altogether, it simplifies evaluation of read mappers and debugging of tools for NGS data.
</p>

<h2>RNFtools</h2>

<p>
	<strong>RNFtools</strong> is an associated software package for RNF,
	which can:
	<ul>
		<li>
			<strong>simulate</strong> RNF-compliant reads using a wide class of integrated read simulators (Art, DwgSim, Mason, WgSim, etc.);
		</li>
		<li>
			<strong>evaluate</strong> mappers using RNF reads;
		</li>
		<li>
			<strong>convert</strong> non-RNF simulated reads to RNF (e.g., from SAM format);
		</li>
		<li>
			<strong>transform</strong> genomic coordinates of RNF reads between different coordinate systems (using chain LiftOver format).
		</li>
	</ul>

	A full documentation with tutorials is available on
	<a href="http://rnftools.rtfd.org">http://rnftools.rtfd.org</a>.

	Source codes of RNFtools are available in its <a href="http://github.com/karel-brinda/rnftools/">GitHub repository</a>.
</p>


<!--
	========================================================
-->
<h2>
	News
</h2>

<table>

<tr>
	<td>
		<strong>21 June 2017</strong>
	</td>
	<td>

		Version 0.3.1. Support for
		<a href="http://rnftools.readthedocs.io/en/latest/tutorial/02_simulation.html#sequence-extraction">sequence extraction</a>.

	</td>
</tr>


<tr>
	<td>
		<strong>18 April 2017</strong>
	</td>
	<td>

		Version 0.3.0. A major update of RNFtools. The SMBL library for
		compiling the used programs on-the-fly is replaced by
		<a href="https://bioconda.github.io/">BioConda</a>.

	</td>
</tr>


<tr>
	<td>
		<strong>3 October 2015</strong>
	</td>
	<td>

		Version 0.2.2. Genome ID for <code>rnftools liftover</code> is
		mandatory.

	</td>
</tr>


<tr>
	<td>
		<strong>17 September 2015</strong>
	</td>
	<td>

		Version 0.2.1. A new subcommand <code>rnftools liftover</code>
		for LiftOvering coordinates in read names.

	</td>
</tr>


<tr>
	<td>
		<strong>31 August 2015</strong>
	</td>
	<td>

		Our paper was  <a
			href="http://dx.doi.org/10.1093/bioinformatics/btv524">accepted
			to Bioinformatics</a>.

	</td>
</tr>

<tr>
	<td>
		<strong>9 July 2015</strong>
	</td>
	<td>
	Version 0.2.0b. List of changes:
	<ul style="">
		<li>A <a href="http://rnftools.readthedocs.org/en/latest/reference/04_command_line.html">command-line variant</a> of RNFtools added (command <code>rnftools</code>).</li>
		<li>Documentation strongly improved (especially the <a href="http://rnftools.readthedocs.org/en/latest/tutorial/00_installation.html">tutorial about installation</a>).</li>
		<li>Many small bugs fixed.</li>
		<li><a href="http://karel-brinda.github.io/rnf-spec">RNF specification</a> slightly updated.</li>
	</ul>
	For a list of all differences, see <a href="https://github.com/karel-brinda/rnftools/pull/14">GitHub diff</a>.
	</td>
</tr>
<tr>
	<td>
	<strong>23 April 2015</strong>
	</td>
	<td>
	Version 0.1.1. Support for Mason in MIShmash.
	</td>
</tr>
<tr>
	<td>
	<strong>19 March 2015</strong>
	</td>
	<td>
	Version 0.1.0. First release of RNFtools.
	</td>
</tr>
</table>

<!--
	========================================================
-->
<h2>
	Publications and posters
</h2>

<ul>

<li>
	K. Břinda,
	V. Boeva, and
	G. Kucherov.
	<strong>RNF: a general framework to evaluate NGS read mappers</strong>.
	<em>Bioinformatics</em>, doi:<a href="http://dx.doi.org/10.1093/bioinformatics/btv524">10.1093/bioinformatics/btv524</a>, 2015.
</li>
<li>
	<!--
		<a href="http://brinda.cz">K. Břinda</a>,
		<a href="https://sites.google.com/site/valentinaboeva/">V. Boeva</a>,
		<a href="http://www-igm.univ-mlv.fr/~koutcher/">G. Kucherov</a>.
	-->
	K. Břinda,
	V. Boeva, and
	G. Kucherov.
	<strong>RNF: a method and tools to evaluate NGS read mappers</strong>.
	<a href="http://brinda.cz/publications/2015_RECOMB_RNF.pdf">Poster</a> on <a href="http://www.recomb2015.pl">RECOMB 2015</a>.

</ul>


<!--
	========================================================
-->

<h2>
<a id="contact" class="anchor" href="#contact" aria-hidden="true"><span class="octicon octicon-link"></span></a>Contact</h2>

<ul>
<li><strong>Karel Břinda</strong> (<a href="http://brinda.cz">website</a>, <a href="mailto:karel.brinda@univ-mlv.fr">e-mail</a>)</li>
<li>Valentina Boeva (<a href="https://sites.google.com/site/valentinaboeva/">website</a>, <a href="mailto:valentina.boeva@curie.fr">e-mail</a>)</li>
<li>Gregory Kucherov (<a href="http://www-igm.univ-mlv.fr/~koutcher/">website</a>, <a href="mailto:gregory.kucherov@univ-mlv.fr">e-mail</a>)</li>
</ul>

		</section>
		<footer>
			<a href="http://ligm.u-pem.fr/">
				<img src="http://ligm.u-pem.fr/uploads/RTEmagicC_LIGM-200.png.png" />
			</a>
			<a href="http://www.curie.fr">
				<img src="http://curie.fr/sites/default/files/logo_fr.png" />
			</a>
			<!--
			<p><small>Hosted on <a href="https://pages.github.com">GitHub Pages</a> using the Dinky theme</small></p>
			-->
		</footer>
	</div>
	<!--[if !IE]><script>fixScale(document);</script><![endif]-->


	</body>
</html>
